base/shell/cmd/misc.c

```diff
@@ -83,45 +83,56 @@ cgetchar (VOID)
 /*
  * Takes a path in and returns it with the correct case of the letters
  */
-VOID GetPathCase( TCHAR * Path, TCHAR * OutPath)
+VOID GetPathCase(IN LPCTSTR Path, OUT LPTSTR OutPath)
 {
-    UINT i = 0;
+    UINT i, cchPath = lstrlen(Path);
     TCHAR TempPath[MAX_PATH];
+    LPTSTR pchTemp = TempPath, pchTempEnd = TempPath + _countof(TempPath);
     WIN32_FIND_DATA FindFileData;
     HANDLE hFind;
-    _tcscpy(TempPath, _T(""));
-    _tcscpy(OutPath, _T(""));
 
-    for(i = 0; i < _tcslen(Path); i++)
+    *pchTemp = OutPath[0] = 0;
+
+    for (i = 0; i < cchPath; ++i)
     {
-        if (Path[i] != _T('\\'))
+        if (pchTemp + 1 >= pchTempEnd)
         {
-            _tcsncat(TempPath, &Path[i], 1);
-            if (i != _tcslen(Path) - 1)
+            StringCchCopy(OutPath, MAX_PATH, Path);
```

Comment:
Why doing a straight copy to OutPath in that case?

Reply:
This is a failed case. To display the failed wildcard pattern, the original text is better.

Reply:
```suggestion
            // If it fails, copy the original path for an error message
            StringCchCopy(OutPath, MAX_PATH, Path);
```